interpreter/cling/lib/Interpreter/ExternalInterpreterSource.cpp

```diff
@@ -0,0 +1,302 @@
+//--------------------------------------------------------------------*- C++ -*-
+// CLING - the C++ LLVM-based InterpreterG :)
+// author:  Elisavet Sakellari <elisavet.sakellari@cern.ch>
+//
+// This file is dual-licensed: you can choose to license it under the University
+// of Illinois Open Source License or the GNU Lesser General Public License. See
+// LICENSE.TXT for details.
+//------------------------------------------------------------------------------
+
+#include "ExternalInterpreterSource.h"
+#include "cling/Interpreter/Interpreter.h"
+
+#include "clang/Lex/Preprocessor.h"
+
+using namespace clang;
+
+namespace {
+  class ClingASTImporter : public ASTImporter {
+  private:
+  cling::ExternalInterpreterSource& m_Source;
+
+  public:
+    ClingASTImporter(ASTContext &ToContext, FileManager &ToFileManager,
+                     ASTContext &FromContext, FileManager &FromFileManager,
+                     bool MinimalImport, cling::ExternalInterpreterSource& source):
+      ASTImporter(ToContext, ToFileManager, FromContext, FromFileManager,
+                  MinimalImport), m_Source(source) {}
+
+    Decl *Imported(Decl *From, Decl *To) override {
+      ASTImporter::Imported(From, To);
+
+      if (isa<TagDecl>(From)) {
+        clang::TagDecl* toTagDecl = dyn_cast<TagDecl>(To);
+        toTagDecl->setHasExternalLexicalStorage();
+        toTagDecl->setMustBuildLookupTable();
+      }
+
+      if (isa<NamespaceDecl>(From))
+      {
+          NamespaceDecl *toNamespaceDecl = dyn_cast<NamespaceDecl>(To);
+          toNamespaceDecl->setHasExternalVisibleStorage();
+      }
+     
+      if (isa<ObjCContainerDecl>(From))
+      {
+          ObjCContainerDecl *toContainerDecl = dyn_cast<ObjCContainerDecl>(To);
+          toContainerDecl->setHasExternalLexicalStorage();
+          toContainerDecl->setHasExternalVisibleStorage();
+      }
+
+      // Put the name of the Decl imported with the
+      // DeclarationName coming from the parent, in  my map.
+      if (NamedDecl *fromNamedDecl = llvm::dyn_cast<NamedDecl>(From)) {
+        NamedDecl *toNamedDecl = llvm::dyn_cast<NamedDecl>(To);
+        m_Source.addToImportedDecls(toNamedDecl->getDeclName(),
+                                    fromNamedDecl->getDeclName());
+      }
+      if (DeclContext *fromDeclContext = llvm::dyn_cast<DeclContext>(From)) {
+        DeclContext *toDeclContext = llvm::dyn_cast<DeclContext>(To);
+        m_Source.addToImportedDeclContexts(toDeclContext, fromDeclContext);
+      }
+
+      return To;
+    }
+  };
+}
+
+namespace cling {
+
+  ExternalInterpreterSource::ExternalInterpreterSource(const cling::Interpreter *parent,
+                                              cling::Interpreter *child) :
+    m_ParentInterpreter(parent), m_ChildInterpreter(child) {
+
+    clang::DeclContext *parentTUDeclContext =
+      clang::TranslationUnitDecl::castToDeclContext(
+      m_ParentInterpreter->getCI()->getASTContext().getTranslationUnitDecl());
```

Comment:
`clang::TranslationUnitDecl` inherits from `DeclContext`, so the conversion happens implicitly: you can just assign it to a `clang::DeclContext*` without any cast.